homeassistant/components/automation/__init__.py

```diff
@@ -244,14 +245,17 @@ async def trigger_service_handler(entity, service_call):
         "async_turn_off",
     )
 
+    reload_lock = asyncio.Lock()
+
     async def reload_service_handler(service_call):
         """Remove all automations and load new ones from config."""
-        conf = await component.async_prepare_reload()
-        if conf is None:
-            return
-        async_get_blueprints(hass).async_reset_cache()
-        await _async_process_config(hass, conf, component)
-        hass.bus.async_fire(EVENT_AUTOMATION_RELOADED, context=service_call.context)
+        async with reload_lock:
```

Comment:
Other integrations are likely to have the same problem.
Instead of adding a lock here, should we add functionality to the service helper to prevent parallel service calls?

Reply:
With this set-up, if someone accidentally runs 3 reloads, we will reload 3 times in a row 🤷 

I would expect one of the following:
 - Only reload once and 2nd + 3rd call will return right away
 - Only reload once and 2nd + 3rd call will return when first run finishes
 - Reload twice. Any call that comes in while a run is active will schedule a run at the end.

Reply:
> With this set-up, if someone accidentally runs 3 reloads, we will reload 3 times in a row 🤷

Yes, but we *have* been asked to reload thrice..

> Only reload once and 2nd + 3rd call will return right away

Seems non-ideal, we can't then guarantee that an automation is reloaded after it was saved.

> Only reload once and 2nd + 3rd call will return when first run finishes

Seems non-ideal, we can't then guarantee that an automation is reloaded after it was saved.

> Reload twice. Any call that comes in while a run is active will schedule a run at the end.

This means functionality is correct, but isn't it making things too complicated.